PWGHF/Tasks/taskMcEfficiency.cxx

```diff
@@ -0,0 +1,290 @@
+    hCandidates->GetAxis(1)->SetTitle("m_{inv}");
+    hCandidates->GetAxis(2)->SetTitle("PDG code");
+    hCandidates->GetAxis(3)->SetTitle("CPA");
+
+    registry.add("hDuplicateCount", "Duplicate count;frequency;count", {HistType::kTH1F, {{10, 0.5, 10.5}}});
+
+    registry.add("hTrackablePtEta", "Prongs kinematics at different steps", {HistType::kStepTHnF, {{20, 0, 10}, {40, -2, 2}}, kNTrackableSteps});
+    auto hTrackablePtEta = registry.get<StepTHn>(HIST("hCandidates"));
+    hTrackablePtEta->GetAxis(0)->SetTitle("p_{T} (GeV/c)");
+    hTrackablePtEta->GetAxis(1)->SetTitle("#eta");
+  }
+
+  template <typename T>
+  inline bool checkTrack(T track)
+  {
+    // TODO configurable?
+    return track.isGlobalTrackWoDCA();
+  }
+
+  template <bool mc, typename T1, typename T2, typename T3>
+  void candidateLoop(T1& candidates, T2& tracks, T3& mcParticles)
+  {
+    using TracksType = std::decay_t<decltype(tracks)>;
+
+    auto hCandidates = registry.get<StepTHn>(HIST("hCandidates"));
+
+    for (auto pdgCode : (std::vector<int>)pdgCodes) {
+      auto decayType = -1;
+      std::array<int, 2> pdgDaughters;
+
+      if (pdgCode == pdg::kD0) {
+        decayType = 1 << DecayType::D0ToPiK;
+        pdgDaughters[0] = +kPiPlus;
+        pdgDaughters[1] = -kKPlus;
+      } else if (pdgCode == pdg::kD0Bar) {
+        decayType = 1 << DecayType::D0ToPiK;
+        pdgDaughters[0] = -kPiPlus;
+        pdgDaughters[1] = +kKPlus;
+      } else {
+        LOGP(fatal, "Not implemented for PDG {}", pdgCode);
+      }
+
+      std::map<int64_t, int> duplicates;
+
+      for (auto& candidate : selectedD0Candidates) {
+        if (!(candidate.hfflag() & decayType)) {
+          continue;
+        }
+
+        bool collisionMatched = false;
+        if constexpr (mc) {
+          auto trackPos = candidate.prong0_as<TracksType>();
+          auto trackNeg = candidate.prong1_as<TracksType>();
+
+          int8_t sign = 0;
+          auto indexRec = RecoDecay::getMatchedMCRec(mcParticles, std::array{trackPos, trackNeg}, pdgCode, pdgDaughters, false, &sign);
+          if (indexRec < 0) {
+            continue;
+          }
+
+          collisionMatched = candidate.collision_as<aod::McCollisionLabels>().mcCollisionId() == mcParticles.iteratorAt(indexRec).mcCollisionId();
+        }
+
+        float mass = -1;
+        bool selected = false;
+        if (pdgCode == pdg::kD0) {
+          mass = invMassD0ToPiK(candidate);
+          selected = candidate.isSelD0() >= selectionFlagD0;
+        } else if (pdgCode == pdg::kD0Bar) {
+          mass = invMassD0barToKPi(candidate);
+          selected = candidate.isSelD0bar() >= selectionFlagD0bar;
+        }
+        LOGP(debug, "Candidate {} has prong {} and prong {} and pT {} and mass {}", candidate.globalIndex(), candidate.prong0Id(), candidate.prong1Id(), candidate.pt(), mass);
+
+        // all candidates
+        hCandidates->Fill(kHFStepTracked, candidate.pt(), mass, pdgCode, candidate.cpa(), collisionMatched);
```

Comment:
`pt` and `cpa` are dynamic columns. Is there a CPU overhead in calling them multiple times or is the compiler smart enough to realise that the return values don't change in between the calls?